src/main/java/uk/gov/hmcts/dts/fact/services/admin/AdminCourtAddressService.java

```diff
@@ -55,7 +65,57 @@ public AdminCourtAddressService(final CourtRepository courtRepository, final Cou
             .collect(toList());
     }
 
-    @SuppressWarnings("PMD.DataflowAnomalyAnalysis")
+    public List<String> validateCourtAddressPostcodes(final String slug, final List<CourtAddress> courtAddresses) {
+        final List<String> invalidPostcodes = new ArrayList<>();
+
+        if (!CollectionUtils.isEmpty(courtAddresses)) {
+            final List<String> allPostcodes = getAllPostcodesSortedByAddressType(courtAddresses);
+
+            if (!CollectionUtils.isEmpty(allPostcodes)) {
+                invalidPostcodes.addAll(validationService.validateFullPostcodes(allPostcodes));
+            }
+
+            // Only update the in-person court coordinates is all input postcodes are valid
+            if (invalidPostcodes.isEmpty()
+                && !CollectionUtils.isEmpty(allPostcodes)
+                && isInPersonCourt(slug)) {
+                updateCourtLatLonUsingPrimaryPostcode(slug, allPostcodes.get(0));
+            }
+        }
+        return invalidPostcodes;
+    }
+
+    private List<String> getAllPostcodesSortedByAddressType(final List<CourtAddress> courtAddresses) {
+        final Map<Integer, uk.gov.hmcts.dts.fact.entity.AddressType> addressTypeMap = addressTypeService.getAddressTypeMap();
+
+        // Sort the postcodes so that the 'visit us' or 'visit or contact us' postcodes appear first
+        return courtAddresses.stream()
+            .sorted(Comparator.comparingInt(a -> AddressType.isCourtAddress(getAddressTypeFromId(addressTypeMap, a.getAddressTypeId())) ? 0 : 1))
+            .map(CourtAddress::getPostcode)
+            .filter(StringUtils::isNotBlank)
+            .collect(toList());
+    }
+
+    private String getAddressTypeFromId(final Map<Integer, uk.gov.hmcts.dts.fact.entity.AddressType> map, final Integer addressTypeId) {
+        if (!map.containsKey(addressTypeId)) {
+            throw new IllegalArgumentException("Unknown address type ID: " + addressTypeId);
+        }
+        return map.get(addressTypeId).getName();
+    }
+
+    private boolean isInPersonCourt(final String slug) {
+        final Court courtEntity = courtRepository.findBySlug(slug)
+            .orElseThrow(() -> new NotFoundException(slug));
+        return courtEntity.getInPerson() == null || courtEntity.getInPerson().getIsInPerson();
+    }
+
+    private void updateCourtLatLonUsingPrimaryPostcode(final String slug, final String postcode) {
+        final Optional<MapitData> mapitData = mapitService.getMapitData(postcode);
+        if (mapitData.isPresent()) {
+            adminService.updateCourtLatLon(slug, mapitData.get().getLat(), mapitData.get().getLon());
+        }
```

Comment:
If we have no mapit data, this assumes that the call to the service returned with an empty result, is that a cause for concern or not? 

Maybe we can provide a warning or error log below, or think about a custom exception in that event